rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,480 @@
+    [Treasury]--> StakingReward: Transfer LP, PBLO\nor other token rewards
+}
+----
+
+TODO: What to do for part of protocol fees that should be transferred to treasury eventually as treasury does not stake it's PBLO?
+
+=== Pallet-Pablo
+
+In order to 1. support LP staking 2. LP trading fee distribution and 3. PBLO staking reward using trading fees, following changes are proposed for https://github.com/ComposableFi/composable/tree/main/frame/pablo[Pallet-Pablo].
+
+[#_feeconfig]
+==== FeeConfig
+
+Each pool in Pablo defines a fee percentage to be charged for each trade.Except for LBPs other pools also define an owner fee that is a percentage out of the main trading fee. The `FeeConfig` is a new abstraction over all fees that could be charged on a pool to allow for extension. At this time a 100% of the owner fee should be defined as a new field `protocol_fee`.
+
+[plantuml,images/pablo-fee-config,png]
+----
+class FeeConfig {
+    +fee_rate: Permill
+    +owner_fee_rate: Permill,
+    +protocol_fee_rate: Permill,
+}
+----
+
+*Existing code must be modified to use this data structure*.
+
+Given this,
+----
+fee = // calculation depends on the pool type: based on the fee_rate
+owner_fee = fee * owner_fee_rate * (1 - protocol_fee_rate);
+protocol_fee = owner_fee * protocol_fee_rate;
+----
+For all pools launched at the Picasso launch following values would be set for these configs
+----
+owner_fee_rate = 20%
+protocol_fee_rate = 100% // all owner fees goes to composable to be distributed as rewards
+----
+
+==== LP Trading Fee Distribution
+
+LPs trading fees are calculated and kept as part of the pool liquidity in Pablo. When LPs remove liquidity from the pool the trading fees are automatically redeemed according their pool LP ratio, check https://hackmd.io/@HaydenAdams/HJ9jLsfTz#Fee-Structure[reference]. This results in trading fee share being diluted overtime for smaller pools as follows.
+
+[stem]
+++++
+"After " n " trades and " m " liquidity additions,"
+
+"trading fees" = sum a_n
+
+"total liquidity" = sum l_m
+
+"fees and liquidity returned for an LP amount l" = l / (sum l_m) * (sum l_m + sum a_n)
+
+= l + l / (sum l_m) * sum a_n
+
+"trading fees received" = x = l / (sum l_m) * sum a_n
+
+"When pool size " sum l_m " increases the amount of trading fees received " x " reduces for a particular LP position."
+
+"For large pool sizes of " sum l_m " (steady state) this effect is negligible, hence it's a good enough strategy to distribute fees."
+
+"But if required this effect can be negated by increasing the trading fee by a " delta a_n " while at the same time subtracting it from the total fees paid out already to liquidity providers as follows,"
+
+"New trading fee " = x_(adjusted) = l / (sum l_m) * (sum a_n + delta a_n)
+
+"For " r_(th) " liquidity provider,"
+
+delta a = (sum_(r+1) l_m) / (sum l_r) * sum a_n
+
+= x_(adjusted) = l_(r) / (sum_(r+1) l_m + sum l_r) * (sum a_n + (sum_(r+1) l_m) / (sum l_r) * sum a_n )
```

Comment:
not sure from where `sum l_r` appears in denumerator.

also can you recheck how `x_(adjusted) ` is simplified into `l_r / (sum l_r) * sum a_n`.

it is so similar to `x = l / (sum l_m) * sum a_n` what is difference of `m` and `r`?

Reply:
> not sure from where sum l_r appears in denumerator.

Just a trick to get the math to work by dividing the full sum to the part before `0 to r th` element and `r+1 to m th` element. `m` is the total number of liquidity additions(LPs).

>also can you recheck how x_(adjusted) is simplified into l_r / (sum l_r) * sum a_n.

Yes, it is correct. 

`sum_(r+1) l_m + sum l_r` parts inside and outside the brackets cancel each other.

> it is so similar to x = l / (sum l_m) * sum a_n what is difference of m and r?

`r` is one of the earlier LPs, `m` is the total number of LPs at the end.

Reply:
Note that trading fees optimization is optional though. It's there because I was trying to understand the problems with calculating trading fees for new LPs without taking into account the dilution to earlier LPs